src/transactions.rs

```diff
@@ -12,468 +12,389 @@ use miniscript::{BitcoinSig, Descriptor, MiniscriptKey, Satisfier, ToPublicKey};
+        feebump_input: Option<(FeeBumpPrevout, u32)>,
+        emer_txout: EmergencyTxOut,
+    ) -> UnvaultEmergencyTransaction {
+        UnvaultEmergencyTransaction(if let Some(feebump_input) = feebump_input {
+            create_tx!(
+                [
+                    (unvault_input.0, unvault_input.1),
+                    (feebump_input.0, feebump_input.1)
+                ],
+                [emer_txout]
             )
-            | (
-                &[RevaultPrevout::UnvaultPrevout(_), RevaultPrevout::FeeBumpPrevout(_)],
-                &[RevaultTxOut::EmergencyTxOut(ref emer_txout)],
-            ) => {
-                let inputs = prevouts
-                    .iter()
-                    .map(|prevout| TxIn {
-                        previous_output: match prevout {
-                            RevaultPrevout::VaultPrevout(ref prev)
-                            | RevaultPrevout::UnvaultPrevout(ref prev)
-                            | RevaultPrevout::FeeBumpPrevout(ref prev) => *prev,
-                            _ => unreachable!(),
-                        },
-                        sequence: RBF_SEQUENCE,
-                        ..Default::default()
-                    })
-                    .collect();
-
-                Ok(RevaultTransaction::EmergencyTransaction(Transaction {
-                    version: 2,
-                    lock_time: 0,
-                    input: inputs,
-                    output: vec![emer_txout.clone()],
-                }))
-            }
-            _ => Err(Error::TransactionCreation(format!(
-                "Emergency: prevout(s) ({:?}) or output(s) ({:?}) type mismatch",
-                prevouts, txouts,
-            ))),
-        }
+        } else {
+            create_tx!([(unvault_input.0, unvault_input.1)], [emer_txout])
+        })
     }
+}
 
-    // It's private on purpose: accessing the inner tx should either be done by using the provided
-    // methods (or adding new ones to be reused), or by careful pattern matching.
-    fn inner_tx(&self) -> &Transaction {
-        match *self {
-            RevaultTransaction::VaultTransaction(ref tx)
-            | RevaultTransaction::UnvaultTransaction(ref tx)
-            | RevaultTransaction::SpendTransaction(ref tx)
-            | RevaultTransaction::CancelTransaction(ref tx)
-            | RevaultTransaction::EmergencyTransaction(ref tx)
-            | RevaultTransaction::FeeBumpTransaction(ref tx) => tx,
-        }
+impl_revault_transaction!(
+    SpendTransaction,
+    doc = "The transaction spending the unvaulting transaction, paying to one or multiple \
+    externally-controlled addresses, and possibly to a new vault txo for the change."
+);
+impl SpendTransaction {
+    /// A spend transaction can batch multiple unvault txouts, and may have any number of
+    /// txouts (including, but not restricted to, change).
+    pub fn new(
+        unvault_inputs: &[(UnvaultPrevout, u32)],
+        spend_txouts: Vec<SpendTxOut>,
```

Comment:
Why have an &[] for inputs but a Vec for outputs?

Reply:
Prevouts are `Copy`, while `TxOut` aren't

Reply:
ie i need to clone the txouts if i take a slice, taking a vec is arguably a better API as i can consume it to the